### PR TITLE
Automatic upgrade of harvester node roles

### PR DESCRIPTION
**Pull request by futuretea**
#259 

## User Experience
On the harvester installation UI, you do not need to select a node role when adding nodes to an existing cluster.

## Solution
Harvester adds a controller to automatically upgrade the role of the node according to the node upgrade policy.

## Known issues or limitations
1. This solution relies on the high availability solution of the built-in etcd of k3s.
Document:  https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/
2. 2-copy etcd Cluster failure tolerance
Due to the failure tolerance policy of etcd, If any master node fails, the cluster will be unavailable. We do not recommend that you use a cluster with two master nodes.

### best practices for k8s

Reference:  https://kubernetes.io/zh/docs/setup/best-practices/cluster-large/ 

1. The load of master nodes and the availability of etcd clusters determine the number of masters.
In other words, there are two reasons for adding master nodes: reducing the load on master nodes and high availability of clusters.
2. etcd does not recommend more than 5 copies.

### harvester scenario

harvester, as an HCI software, is deployed in bare metal servers. As the machine specifications of bare metal servers are relatively large in production, we can assume that three Master Nodes can cope with most cluster sizes. The main purpose of adding master nodes can be considered as high availability of clusters. Therefore, nodes should be added as master nodes first to achieve high availability as soon as possible.

To sum up, the following strategies are designed:
1. The first three nodes are added to the cluster as management nodes.
2. Subsequent nodes are used as workers to join the cluster.
3. 5 management nodes scenario is reserved to handle large cluster sizes.

## Impact of node upgrade

The k3s agent includes kubelet and kube-proxy components.
Therefore, unavailability during upgrade will result in:
1. Unable to create a pod.
2. Service access failure.

## Conditions for node upgrade

1. The node is the node of harvester.
The harvester node must be labeled with a default label ~~`harvester.cattle.io/version`~~ `harvester.io/managed` to identify the harvester node.
2. The status of the node is not being upgraded.
Although the node upgrade operation should be implemented as idempotent, repeated operations in the upgraded State should be avoided.
3. node is healthy.

## Upgrade node selection

~~When the total number of nodes in the cluster reaches the threshold of the upgrade policy, node upgrade is triggered.
After excluding nodes that do not meet the node upgrade conditions, upgrade the youngest nodes.
Since the newly added node does not have a user workload, the impact on the user can be minimized.~~ first three nodes

## Node upgrade process
1. Cordon node (disable scheduling) ~~and add svclb label `svccontroller.k3s.cattle.io/enablelb="true"`~~
2. Upgrade
  * copy `/k3os/system/config.yaml`  to `/var/lib/rancher/k3os/config.yaml`
  * change `agent` in `/var/lib/rancher/k3os/config.yaml` to `server`
  * change `agent` in `/etc/init.d/k3s-service` to `server`
  * send SIGHUP to [k3s-restarter](https://raw.githubusercontent.com/futuretea/k3os-vagrant/main/scripts/k3s-restarter-installer.sh) in the host （need `sleep 5`  to wait job complete and restart k3s service from host）
3. Resume scheduling

## Node upgrade method

To modify the configuration file in host, you can use a privileged Job to execute an upgrade script.
Reference： https://github.com/rancher/system-upgrade-controller

## Havester installer Dependency 

1. dump k3os version to 1.19.5.rc.1
2.  add k3s-restarter
3.  add --cluster-init to first server's k3s_args
4. add node label ~~"harvester.cattle.io/version"~~  "harvester.io/managed"

### CODE REVIEW

**Comment by guangbochen** [comment]
Note: we will need to take care of the `serviceLB` node selection on HA mode when the first node is down or removed.
https://rancher.com/docs/k3s/latest/en/networking/#excluding-the-service-lb-from-nodes

File: pkg/controller/master/node/controller.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	statefulSetCache ctlappsv1.StatefulSetCache
+	recorder         record.EventRecorder
+}
+
+// OnNodeChanged automate the upgrade of node roles and balance of minio pods
+// If the number of masters in the cluster is less than spec number,
+// the harvester node with the smallest age is automatically promoted to be master.
+// If minio is deployed in distribute mode, make minio pods balanced.
+// Minio balancing only takes place when the node is ready and not in the upgrade state.
+func (h *Handler) OnNodeChanged(key string, node *corev1.Node) (*corev1.Node, error) {
```
Comment (gitlawr) [inline pkg/controller/master/node/controller.go]: Can we separate the MinIO rebalancing logic and node promotion logic into different controllers?
One controller one task. One controller one file.

File: pkg/controller/master/node/promote_check.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+// If the cluster doesn't need to be promoted, return nil
+func (h *Handler) selectPromoteNodeList(nodeList []*corev1.Node) ([]*corev1.Node, error) {
+	masterNumber := 0
+	ignoreNodeNumber := 0
+	canBePromoteNodeList := make([]*corev1.Node, 0, len(nodeList))
+	for _, node := range nodeList {
+		if h.shouldBeMaster(node) {
+			masterNumber++
+		} else {
+			if h.isHarvesterNode(node) {
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_check.go]: Check earlier before listing all nodes

File: pkg/controller/master/node/controller.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	promoteNodeList, err := h.selectPromoteNodeList(nodeList)
+	if err != nil {
+		return nil, err
+	}
+
+	switch len(promoteNodeList) {
+	case 0:
+		return h.makeMinioBalanced(key, node)
+	case 1:
+		return h.promote(promoteNodeList[0])
```
Comment (gitlawr) [inline pkg/controller/master/node/controller.go]: Looks like this is handled in the default case so we probably don't need case 1.

File: pkg/controller/master/node/promote_job.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	j, err := h.jobs.Create(job)
+	return j, checkError(fmt.Sprintf("create job %s on node %s", job.Name, node.Name), err)
+}
+
+func (h *Handler) deleteJob(job *batchv1.Job, deletionPropagation metav1.DeletionPropagation) error {
+	err := h.jobs.Delete(job.Namespace, job.Name, &metav1.DeleteOptions{PropagationPolicy: &deletionPropagation})
+	return checkError(fmt.Sprintf("delete job %s", job.Name), err)
+}
+
+func buildPromoteJob(nodeName string, image string, command, args []string) *batchv1.Job {
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_job.go]: If image, command, and args are constants, we don't need to pass them as parameters

File: pkg/controller/master/node/controller.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+	if ConditionJobFailed.IsTrue(job) {
+		return h.setPromoteResult(job, node, PromoteStatusFailed)
+	}
+
+	return job, nil
+}
+
+// OnJobRemove
+// If the running promote job is deleted, the node's promote status will be marked as unknown
```
Comment (gitlawr) [inline pkg/controller/master/node/controller.go]: How does it recover from this situation?

**Comment by gitlawr** [comment]
If there's any dependency on the installer, We need to link that here.

File: pkg/controller/master/node/controller.go (hunk 5 above)
Comment (futuretea) [reply pkg/controller/master/node/controller.go]: In this situation, it can only be handled manually.

File: pkg/controller/master/node/promote_check.go (hunk 2 above)
Comment (futuretea) [reply pkg/controller/master/node/promote_check.go]: This logic is take care about user already has some master (not installed by harvester-installer)

File: pkg/controller/master/node/controller.go (hunk 1 above)
Comment (futuretea) [reply pkg/controller/master/node/controller.go]: fixed

**Comment by futuretea** [comment]
> Note: we will need to take care of the `serviceLB` node selection on HA mode when the first node is down or removed.
> https://rancher.com/docs/k3s/latest/en/networking/#excluding-the-service-lb-from-nodes

Each master node will add a svclb label now.

**Comment by futuretea** [comment]
> If there's any dependency on the installer, We need to link that here.

updated

File: pkg/controller/master/node/promote_controller.go [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+
+	PromoteStatusComplete = "complete"
+	PromoteStatusRunning  = "running"
+	PromoteStatusUnknown  = "unknown"
+	PromoteStatusFailed   = "failed"
+
+	defaultSpecMasterNumber = 3
+)
+
+const (
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_controller.go]: combine two const blocks to one

File: pkg/controller/master/node/promote_controller.go [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+	promoteControllerName = "promote-node-controller"
+
+	KubeNodeRoleLabelPrefix = "node-role.kubernetes.io/"
+	KubeMasterNodeLabelKey  = KubeNodeRoleLabelPrefix + "master"
+
+	KubeNodeSVCLabelPrefix      = "svccontroller.k3s.cattle.io/"
+	KubeNodeSVCEnableLBLabelKey = KubeNodeSVCLabelPrefix + "enablelb"
+
+	HarvesterLabelAnnotationPrefix      = "harvester.cattle.io/"
+	HarvesterVersionLabelKey            = HarvesterLabelAnnotationPrefix + "managed"
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_controller.go]: Update the name since it's irrelevant to the version

File: pkg/controller/master/node/promote_controller.go [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+			masterNumber++
+		} else {
+			if h.isHarvesterNode(node) {
+				if h.isHealthyNode(node) {
+					canBePromoteNodeList = append(canBePromoteNodeList, node)
+				}
+			} else {
+				ignoreNodeNumber++
+			}
+		}
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_controller.go]: Looks like nested conditions can be avoided
e.g.
```
if shoudBeMaster{
}else if isHarvesterNode() && isHealthyNode(){
else{
}
```

File: pkg/controller/master/node/promote_controller.go [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+}
+
+func checkError(action string, err error) error {
+	if err != nil {
+		logrus.Errorf("%s failed, %s", action, err)
+		return err
+	}
+	logrus.Infof("%s", action)
+	return nil
+}
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_controller.go]: Some cons of this pattern:
1. We usually return errors to the caller and don't log the error message directly(The caller handles it). Otherwise it may result in duplicate error logs. you can return a wrapped error if you want to provide more info.
2. every call for this generates an info level log, but looks like some implementation details users may not need to know?
3. It is called checkError but seems to be a logging util function.

In short, I would suggest we do the error check and logging directly instead of wrapping them here

File: pkg/controller/master/node/promote_controller.go (hunk 8 above)
Comment (futuretea) [reply pkg/controller/master/node/promote_controller.go]: if node is harvester agent node but unhealthy. should not `ignoreNodeNumber++`

Comment (gitlawr) [reply pkg/controller/master/node/promote_controller.go]: Should we ignore all nodes that cannot be promoted?
Consider the following case:
1 master 1 healthy 1 unhealthy
masterNumber=1, ignoreNodeNumber=0
specMasterNumber=len(nodeList) - ignoreNodeNumber = 3

But given that there are no enough healthy nodes, do we want to do promote in this case?

Comment (futuretea) [reply pkg/controller/master/node/promote_controller.go:279]: In this case, the apiserver will be unavailable, three master node is more stable.

File: pkg/controller/master/node/promote_controller.go [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+		{
+			Name:      "promote",
+			Image:     promoteImage,
+			Command:   []string{"sh"},
+			Args:      []string{"-c", fmt.Sprintf(`chroot %s bash -c "%s"`, promoteRootMountPath, promoteCommand)},
+			Resources: corev1.ResourceRequirements{},
+			VolumeMounts: []corev1.VolumeMount{
+				{Name: "host-root", MountPath: promoteRootMountPath},
+			},
+			ImagePullPolicy: promoteImagePullPolicy,
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_controller.go]: can we use corev1.PullIfNotPresent directly here

File: pkg/controller/master/node/promote_controller.go [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+			SecurityContext: &corev1.SecurityContext{
+				Capabilities: &corev1.Capabilities{
+					Add: []corev1.Capability{
+						"CAP_SYS_BOOT",
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_controller.go]: This doesn't look like a proper capability for the command

File: pkg/controller/master/node/promote_controller.go [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+// selectPromoteNodeList return a list contain nodes need to be promoted
+// If the cluster doesn't need to be promoted, return nil
+func (h *PromoteHandler) selectPromoteNodeList(nodeList []*corev1.Node) ([]*corev1.Node, error) {
+	masterNumber := 0
+	canBePromoteNodeList := make([]*corev1.Node, 0, len(nodeList))
+	for _, node := range nodeList {
+		if h.shouldBeMaster(node) {
+			masterNumber++
+		} else {
+			if h.isHarvesterNode(node) && h.isHealthyNode(node) {
```
Comment (gitlawr) [inline pkg/controller/master/node/promote_controller.go]: else if

**Review by gitlawr — APPROVED** [review]
LGTM

File: pkg/controller/master/node/promote_controller.go [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+func (s SortByAgeNodeList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s SortByAgeNodeList) Less(i, j int) bool {
+	// reverse order
+	return !s[i].CreationTimestamp.Before(&s[j].CreationTimestamp)
+}
+
+// selectPromoteNodeList return a list contain nodes need to be promoted
+// If the cluster doesn't need to be promoted, return nil
+func (h *PromoteHandler) selectPromoteNodeList(nodeList []*corev1.Node) ([]*corev1.Node, error) {
+	masterNumber := 0
```
Comment (guangbochen) [inline pkg/controller/master/node/promote_controller.go]: use better variable semantics `var` here

File: pkg/controller/master/node/promote_controller.go [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+	if promoteNodeNumber <= 0 {
+		return nil, nil
+	}
+
+	// make sure have enough nodes can be promote
+	if len(canBePromoteNodeList) < promoteNodeNumber {
+		return nil, nil
+	}
+
+	// sort by creation time from largest to smallest to find the youngest node that can be promoted
```
Comment (guangbochen) [inline pkg/controller/master/node/promote_controller.go]: `\s\youngest\latest`

File: pkg/controller/master/node/promote_controller.go [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+// If the cluster doesn't need to be promoted, return nil
+func (h *PromoteHandler) selectPromoteNodeList(nodeList []*corev1.Node) ([]*corev1.Node, error) {
+	masterNumber := 0
+	canBePromoteNodeList := make([]*corev1.Node, 0, len(nodeList))
+	for _, node := range nodeList {
+		if h.shouldBeMaster(node) {
+			masterNumber++
+		} else if h.isHarvesterNode(node) && h.isHealthyNode(node) {
+			canBePromoteNodeList = append(canBePromoteNodeList, node)
+		}
```
Comment (guangbochen) [inline pkg/controller/master/node/promote_controller.go]: this is not thread-safe to count the promote node,  for example when adding both the 3rd and 4th node simultaneously via PXE, none of the existing worker nodes has been marked as `promoted-node` yet, and according to the current reconcile logic it will promote 2 groups of the master node(2nd and 3rd, 3rd and 4th) to be the new master node, as a result, we will get a total number of 4 master node.
Tips: add promote label on the early state and check if we can add locker to the promote controller.

**Review by guangbochen — APPROVED** [review]
LGTM
